adc/ad7779/ad7779-tool.c

```diff
@@ -0,0 +1,525 @@
+	memcpy(msg.o.raw, &ioctl, sizeof(adc_dev_ctl_t));
+
+	log_verbose("enable\n");
+	if (msgSend(port, &msg) < 0 || msg.o.io.err != 0) {
+		printf("sample: failed to enable adc: %d\n", msg.o.io.err);
+		return 1;
+	}
+
+	return 0;
+}
+
+
+static uint32_t read_intr(uint32_t port)
+{
+	msg_t msg = { 0 };
+	uint32_t intr = 0;
+
+	msg.type = mtRead;
+	msg.o.size = sizeof(uint32_t);
+	msg.o.data = &intr;
+
+	if (msgSend(port, &msg) || msg.o.io.err < 0) {
+		printf("MSG failed to read adc intr: %s", strerror(msg.o.io.err));
+	}
+
+	return intr;
+}
+
+
+static void signal_handler(int signal)
+{
+	tool_common.running = 0;
+}
+
+
+static inline uint8_t raw2channel(uint32_t raw)
+{
+	return (raw & 0x70000000) >> 28;
+}
+
+
+static inline int32_t raw2value(uint32_t raw)
+{
+	int32_t value = raw & 0x00FFFFFFu;
+	if (value >= 1 << 23)
+		value -= 1 << 24;
+	return value;
+}
+
+
+#define print_us(type_, us_) printf("time " type_ ": %llu.%03llu ms\n", (us_) / 1000, (us_) % 1000);
+
+static int adc_read(uint32_t port, uint32_t readIntCnt, int streamFd, int dumpSamplerate, int dumpAverage)
+{
+	uint32_t first_intr = 0;
+	uint32_t last_intr = 0;
+
+	int64_t sum[TOTAL_CHANNELS] = {};
+	uint32_t sumcnt = 0;
+	int retCode = 0;
+
+
+	tool_common.running = 1;
+
+	struct timespec start, end;
+	for (int i = 1; i <= readIntCnt; ++i) {
+		uint32_t intr = read_intr(port);
+
+		if (first_intr == 0) {
+			clock_gettime(CLOCK_MONOTONIC_RAW, &start);
+			first_intr = intr;
+		}
+
+		if (last_intr != 0 && last_intr + 1 != intr) {
+			printf("missed intr! %u -> %u\n", last_intr, intr);
+			retCode = 1; /* signal with exit code that data might not be reliable */
+		}
+
+		last_intr = intr;
+
+		/* support infinite running until ^C */
+		if (tool_common.running == 0) {
+			break;
+		}
+
+#if 0 /* simulate periodic samplerate monitoring in infinite stream (meter-srv use case) */
+		const int readCntEstimateStep = 100;
+
+		if ((dumpSamplerate != 0) && (i % readCntEstimateStep == 0)) {
+			clock_gettime(CLOCK_MONOTONIC_RAW, &end);
+
+			uint64_t us = (end.tv_sec * 1000000 + end.tv_nsec / 1000) - (start.tv_sec * 1000000 + start.tv_nsec / 1000);
+			uint64_t us_per_buffer = us / (last_intr - first_intr);
+			uint64_t us_per_sample = us_per_buffer / (tool_common.config.size / 8 / sizeof(uint32_t));
+
+			if (tool_common.verbose > 0) {
+				printf("read operations: %u\n", readCntEstimateStep);
+				printf("ints: %u - %u  + 1 = %u\n", last_intr, first_intr, last_intr - first_intr + 1);
+				print_us("total", us);
+				print_us("per buffer", us_per_buffer);
+				print_us("per sample", us_per_sample);
+			}
+
+			printf("freq = %6.3f\n", (float)1000000 / us_per_sample);
+
+			first_intr = 0; /* start new measure window */
+		}
+#endif
+
+		/* dumping & averaging */
+		if (streamFd >= 0 || dumpAverage != 0) {
+			volatile uint32_t *samplebuf = tool_common.config.ptr[(intr - 1) % tool_common.config.nr];
+
+			const int frames = tool_common.config.size / sizeof(uint32_t) / TOTAL_CHANNELS;
```

Comment:
```suggestion
			const int frames = (tool_common.config.size / sizeof(uint32_t)) / TOTAL_CHANNELS;
```

Reply:
not needed (operator `/` has left-to-right associativity per standard), but added for readability?